tower-http/src/request_id.rs

```diff
@@ -0,0 +1,575 @@
+    type Service = PropagateRequestId<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        PropagateRequestId::new(inner, self.header_name.clone())
+    }
+}
+
+/// Propagate request ids from requests to responses.
+///
+/// See the [module docs](self) for an example.
+///
+/// If the request contains a matching header that header will be applied to responses. If a
+/// [`RequestId`] extension is also present it will be propagated as well.
+#[derive(Debug, Clone)]
+pub struct PropagateRequestId<S> {
+    inner: S,
+    header_name: HeaderName,
+}
+
+impl<S> PropagateRequestId<S> {
+    /// Create a new `PropagateRequestId`.
+    pub fn new(inner: S, header_name: HeaderName) -> Self {
+        Self { inner, header_name }
+    }
+
+    /// Create a new `PropagateRequestId` that uses `x-request-id` as the header name.
+    pub fn x_request_id(inner: S) -> Self {
+        Self::new(inner, HeaderName::from_static(X_REQUEST_ID))
+    }
+
+    define_inner_service_accessors!();
+
+    /// Returns a new [`Layer`] that wraps services with a `PropagateRequestId` middleware.
+    pub fn layer(header_name: HeaderName) -> PropagateRequestIdLayer {
+        PropagateRequestIdLayer::new(header_name)
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for PropagateRequestId<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = PropagateRequestIdResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let request_id = req
+            .headers()
+            .get(&self.header_name)
+            .cloned()
+            .map(RequestId::new);
+
+        PropagateRequestIdResponseFuture {
+            inner: self.inner.call(req),
+            header_name: self.header_name.clone(),
+            request_id,
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`PropagateRequestId`].
+    pub struct PropagateRequestIdResponseFuture<F> {
+        #[pin]
+        inner: F,
+        header_name: HeaderName,
+        request_id: Option<RequestId>,
+    }
+}
+
+impl<F, B, E> Future for PropagateRequestIdResponseFuture<F>
+where
+    F: Future<Output = Result<Response<B>, E>>,
+{
+    type Output = Result<Response<B>, E>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let mut response = futures_core::ready!(this.inner.poll(cx))?;
+
+        if let Some(current_id) = response.headers().get(&*this.header_name) {
+            if response.extensions().get::<RequestId>().is_none() {
+                let current_id = current_id.clone();
+                response.extensions_mut().insert(current_id);
+            }
+        } else if let Some(request_id) = this.request_id.take() {
```

Comment:
it shouldn't be `else if`
It should `if let` as you definitely want to propagate request id if it is set

Reply:
Its an `else if` because If the response already has a request id for some reason, perhaps another middleware that conditionally applies it, then I don't think we should override it.

Reply:
that sounds like a programming error to me
Well, it is up to you, but I think user shouldn't have multiple layers that set request id

Reply:
I don't think it hurts to support.

Reply:
I mean I'm just speculating so maybe not 🤷‍♂️ Would like to hear what others think.

Reply:
I agree that I don't think it hurts to support.